Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,905 @@
+        } else {
+            code.m_symbol_codes.extend({ 0b100, 0b101, 0b110, 0b111 });
+            quick_sort(symbols);
+            code.m_symbol_values = move(symbols);
+        }
+    }
+
+    return {};
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_complex_prefix_code(CanonicalCode& code, size_t alphabet_size, size_t hskip)
+{
+    // hskip should only be 0, 2 or 3
+    VERIFY(hskip != 1);
+    VERIFY(hskip <= 3);
+
+    // Read the prefix code_value that is used to encode the actual prefix code_value
+    size_t const symbol_mapping[18] = { 1, 2, 3, 4, 0, 5, 17, 6, 16, 7, 8, 9, 10, 11, 12, 13, 14, 15 };
+    size_t code_length[18] { 0 };
+    size_t code_length_counts[6] { 0 };
+
+    size_t sum = 0;
+    size_t number_of_non_zero_symbols = 0;
+    for (size_t i = hskip; i < 18; i++) {
+        size_t len = TRY(read_complex_prefix_code_length());
+        code_length[symbol_mapping[i]] = len;
+
+        if (len != 0) {
+            code_length_counts[len]++;
+            sum += (32 >> len);
+            number_of_non_zero_symbols++;
+        }
+
+        if (sum == 32)
+            break;
+        VERIFY(sum < 32);
+    }
+
+    BrotliDecompressionStream::CanonicalCode temp_code;
+    if (number_of_non_zero_symbols > 1) {
+        size_t code_value = 0;
+        for (size_t bits = 1; bits <= 5; bits++) {
+            code_value = (code_value + code_length_counts[bits - 1]) << 1;
+            size_t current_code_value = code_value;
+
+            for (size_t i = 0; i < 18; i++) {
+                size_t len = code_length[i];
+                if (len == bits) {
+                    temp_code.m_symbol_codes.append((1 << bits) | current_code_value);
+                    temp_code.m_symbol_values.append(i);
+                    current_code_value++;
+                }
+            }
+        }
+    } else {
+        for (size_t i = 0; i < 18; i++) {
+            size_t len = code_length[i];
+            if (len != 0) {
+                temp_code.m_symbol_codes.append(1);
+                temp_code.m_symbol_values.append(i);
+                break;
+            }
+        }
+    }
+
+    // Read the actual prefix code_value
+    sum = 0;
+    size_t i = 0;
+
+    size_t previous_non_zero_code_length = 8;
+    size_t last_symbol = 0;
+    size_t last_repeat = 0;
+
+    Vector<size_t> result_symbols;
+    Vector<size_t> result_lengths;
+    size_t result_lengths_count[16] { 0 };
+    while (i < alphabet_size) {
+        VERIFY(sum < 32768);
+        auto symbol = temp_code.read_symbol(m_input_stream).release_value();
+
+        if (symbol < 16) {
+            result_symbols.append(i);
+            result_lengths.append(symbol);
+            result_lengths_count[symbol]++;
+
+            i++;
+
+            if (symbol != 0) {
+                previous_non_zero_code_length = symbol;
+                sum += (32768 >> symbol);
+                if (sum == 32768)
+                    break;
+                VERIFY(sum < 32768);
```

Comment:
This can be reached with: [crash-f387b948147677a78d1709903863eac40ab88a78.txt](https://github.com/SerenityOS/serenity/files/8421915/crash-f387b948147677a78d1709903863eac40ab88a78.txt)